src/services/ProveGatewayService.ts

```diff
@@ -0,0 +1,181 @@
+import BigNumber from 'bignumber.js';
+import { GatewayRepository } from '../repositories/GatewayRepository';
+import { MessageRepository } from '../repositories/MessageRepository';
+import Logger from '../Logger';
+import Utils from '../Utils';
+import { AUXILIARY_GAS_PRICE } from '../Constants';
+import Observer from '../observer/Observer';
+import { AuxiliaryChain } from '../repositories/AuxiliaryChainRepository';
+
+const Mosaic = require('@openst/mosaic.js');
+
+export default class ProveGatewayService extends Observer<AuxiliaryChain> {
+  private gatewayRepository: GatewayRepository;
+
+  private messageRepository: MessageRepository;
+
+  private originWeb3: object;
+
+  private auxiliaryWeb3: object;
+
+  private auxiliaryWorkerAddress: string;
+
+  private gatewayAddress: string;
+
+  private auxiliaryChainId: number;
+
+  /**
+   *  Constructor
+   *
+   * @param gatewayRepository Instance of auxiliary chain repository.
+   * @param messageRepository Instance of message repository.
+   * @param originWeb3 Origin Web3 instance.
+   * @param auxiliaryWeb3 Auxiliary Web3 instance.
+   * @param auxiliaryWorkerAddress auxiliary worker address, this should be
+   *                               unlocked and added in web3 wallet.
+   * @param gatewayAddress Address of gateway contract on origin chain.
+   * @param auxiliaryChainId Auxiliary chain Id.
+   */
+  public constructor(
+    gatewayRepository: GatewayRepository,
+    messageRepository: MessageRepository,
+    originWeb3: object,
+    auxiliaryWeb3: object,
+    auxiliaryWorkerAddress: string,
+    gatewayAddress: string,
+    auxiliaryChainId: number,
+  ) {
+    super();
+
+    this.gatewayRepository = gatewayRepository;
+    this.originWeb3 = originWeb3;
+    this.auxiliaryWeb3 = auxiliaryWeb3;
+    this.auxiliaryWorkerAddress = auxiliaryWorkerAddress;
+    this.gatewayAddress = gatewayAddress;
+    this.messageRepository = messageRepository;
+    this.auxiliaryChainId = auxiliaryChainId;
+
+    // Suppressing lint error
+    // Refer: https://github.com/typescript-eslint/typescript-eslint/issues/636
+    /* eslint-disable @typescript-eslint/unbound-method */
+    this.reactTo = this.reactTo.bind(this);
```

Comment:
I am not sure why it's needed to bind function to itself?

Reply:
Because of scope change due to arrow functions.

Reply:
Could you please give me more details about this? I am not sure what arrow functions are we referring here? Should we do it in all classes we have?

Reply:
Yes, In mosaic.js it has been followed to bind functions to this. 
https://github.com/OpenST/mosaic.js/blob/e7326c0326822ae01e2fc744468a1208feaeeb37/src/ContractInteract/EIP20CoGateway.js#L67

This article may explain bind usage(I just quickly googled it) 
https://www.smashingmagazine.com/2014/01/understanding-javascript-function-prototype-bind/

Reply:
It works after removing bind. I have removed the binding. I will read more about this.